### PR TITLE
Add search data store

### LINKED ISSUE #824: Hibernate Search Data Store
Elide should provide a DataStore that can delegate a subset of full text search queries to [Hibernate Search](https://hibernate.org/search/).   This would improve performance for text search use cases via Elide. 

## Expected Behavior

Hibernate Search couples capabilities from Hibernate's ORM with Lucene (on-box) or Elasticsearch (clustered) to get the best of both worlds:

- Acid Transactions, Joins, RDBMS features from the ORM
- Full text search with Lucene/Elasticsearch.

Hibernate Search provides a set of annotations that decorate a JPA entity and inform Hibernate how to index a given entity in Lucene.  The index and the database are kept in sync whenever an entity is modified via Hibernate.

## Current Behavior
Elide supports text search currently by crafting JPQL queries in the form:
```SELECT id, field1, field2, ... FROM entity WHERE field1 like '%search string%'```

This is not ideal because:

- Databases cannot leverage b-tree indices when a search term starts with a wildcard.
- Case insensitive searches may require changing the case of the search field in the query (which can bypass any index on that column).
- These searches typically involve full table scans - which won't perform for larger tables.

## Possible Solution

Elide can create a new `HibernateSearchStore` that can wrap another `EntityManager` based store (JPA, Hibernate5, etc).

The store normally delegates all operations to the wrapped store.  However, if the filter expression for `loadObjects` or `getRelation` only contains attributes of an entity that is:

- Indexed in Lucene (the model is marked with the `@Indexed` annotation)
- The filter predicates only reference fields that are indexed in Lucene (the fields are marked with `@Field`)

the data store will instead construct a HibernateSearch query against Lucene/elasticsearch.

Queries would be constructed using the simple query string builder:

```java
Query query = lucene.simpleQueryString()
    .onFields("title")
    .matching("war + (peace | harmony)")
    .createQuery();
```
The following Elide operators would be supported (all others would result in an Exception being thrown):

1. IN - would map to a phrase query
2. INFIX - would map to a keyword prefix query
2. PREFIX - would map to a keyword prefix query
4. NOT - would map to NOT boolean operator

### PR DESCRIPTION
**Pull request by aklish**
Resolves #824
## Description

Alpha version of full text search for Elide.

## Requirements

This store leverages [Hibernate Search](https://hibernate.org/search/) which requires Hibernate 5.4+.

## Usage

`SearchDataStore` wraps another fully featured store and supports full text search on fields that are indexed using Hibernate Search.
If the query cannot be answered by the `SearchDataStore`, it delegates the query to the underlying (wrapped) data store.

### Annotate Your Entity 

Use Hibernate Search annotations to describe how your entities are indexed and stored in Lucene or Elasticsearch.

```java
@Entity
@Include
@Indexed
@Data
@AnalyzerDef(name = "case_sensitive",
        tokenizer = @TokenizerDef(factory = StandardTokenizerFactory.class)
)
@AnalyzerDef(name = "case_insensitive",
        tokenizer = @TokenizerDef(factory = WhitespaceTokenizerFactory.class),
        filters = {
                @TokenFilterDef(factory = LowerCaseFilterFactory.class)
        }
)
public class Item {
    @Id
    private long id;

    @Fields({
            @Field(name = "name", index = Index.YES,
                    analyze = Analyze.YES, store = Store.NO, analyzer = @Analyzer(definition = "case_insensitive")),
            @Field(name = "sortName", analyze = Analyze.NO, store = Store.NO, index = Index.YES)
    })
    @SortableField(forField = "sortName")
    private String name;

    @Field(index = Index.YES, analyze = Analyze.YES,
            store = Store.NO, analyzer = @Analyzer(definition = "case_insensitive"))
    private String description;

    @Field(index = Index.YES, analyze = Analyze.NO, store = Store.NO)
    @DateBridge(resolution = Resolution.MINUTE, encoding = EncodingType.STRING)
    @SortableField
    private Date modifiedDate;

    private BigDecimal price;
}
```
### Wrap Your DataStore

```java

/* Create your ORM based data store */
DataStore store = ...

/* Wrap it with a SearchDataStore */
EntityManagerFactory emf = Persistence.createEntityManagerFactory("MyPersistenceUnitName");

searchStore = new SearchDataStore(mockStore, emf, true);

/* Configure Elide with your store */
ElideSettings = new ElideSettingsBuidler(searchStore).build();
```

### Indexing your Data
You can index data either by:
1. When the `SearchDataStore` is initalized, indicate (by setting `indexOnStartup` to true) that it should build a complete index.
2. Issuing created, updated, and delete requests against your Elide service.
3. Using an out of band process using Hibernate Search APIs.

## Caveats
1. Only text fields (String) are supported/tested.  Other data types (dates, numbers, etc) have not been tested.
2. Only IN, NOT, INFIX, and PREFIX operators (and their case insensitive equivalents) are supported.

## Motivation and Context
RDBMS wildcard searches require full table scans for the INFIX operator.  For large dimensions, this does not perform.

## How Has This Been Tested?
Unit tests with an H2 database that is used to populate an on-box lucene index.

## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: elide-datastore/elide-datastore-search/src/main/java/com/yahoo/elide/datastores/search/SearchDataTransaction.java [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+     * @param expression An elide filter expression.
+     * @param entityClass The entity being searched.
+     * @return true if Lucene can be used.  False otherwise.
+     */
+    private boolean canSearch(FilterExpression expression, Class<?> entityClass) {
+        Collection<FilterPredicate> predicates = expression.accept(new PredicateExtractionVisitor());
+
+        return predicates.stream().allMatch((predicate) -> {
+            if (predicate.getPath().getPathElements().size() != 1) {
+                return false;
```
Comment (QubitPi) [inline elide-datastore/elide-datastore-search/src/main/java/com/yahoo/elide/datastores/search/SearchDataTransaction.java]: What would make a multi-entity path unsearchable?

Comment (aklish) [reply elide-datastore/elide-datastore-search/src/main/java/com/yahoo/elide/datastores/search/SearchDataTransaction.java]: Multiple path elements means we have to join to other tables.

File: elide-datastore/elide-datastore-search/src/main/java/com/yahoo/elide/datastores/search/FilterExpressionToLuceneQuery.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            case INFIX_CASE_INSENSITIVE:
+            case PREFIX_CASE_INSENSITIVE:
+            case PREFIX:
+            case INFIX: {
+                return predicateValues.stream()
+                        .map((str) -> str + "*")
+                        .collect(Collectors.joining(" | "));
+            }
+            default:
+                throw new IllegalArgumentException("Unsupported Predicate Operator: " + op);
```
[the remark this answers is not in the document]
Comment (aklish) [reply elide-datastore/elide-datastore-search/src/main/java/com/yahoo/elide/datastores/search/FilterExpressionToLuceneQuery.java]: Maybe - but not in this version.

File: elide-datastore/elide-datastore-search/README.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+### Annotate Your Entity 
+
+Use Hibernate Search annotations to describe how your entities are indexed and stored in Lucene or Elasticsearch.
+
+```java
+@Entity
+@Include
+@Indexed
+@Data
```
Comment (jkusa) [inline elide-datastore/elide-datastore-search/README.md]: is this hibernate search or lombok?

File: elide-datastore/elide-datastore-search/README.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+### Wrap Your DataStore
+
+```java
+/* Create your ORM based data store */
+DataStore store = ...
+
+/* Wrap it with a SearchDataStore */
+EntityManagerFactory emf = Persistence.createEntityManagerFactory("MyPersistenceUnitName");
+
+searchStore = new SearchDataStore(mockStore, emf, true);
```
Comment (jkusa) [inline elide-datastore/elide-datastore-search/README.md]: Can we comment the last param?

File: elide-datastore/elide-datastore-search/src/main/java/com/yahoo/elide/datastores/search/FilterExpressionToLuceneQuery.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    public Query visitAndExpression(AndFilterExpression expression) {
+        return builder.bool()
+                .must(expression.getLeft().accept(this))
+                .must(expression.getRight().accept(this))
+                .createQuery();
+    }
+
+    @Override
+    public Query visitOrExpression(OrFilterExpression expression) {
+        Query query = builder.bool()
```
Comment (jkusa) [inline elide-datastore/elide-datastore-search/src/main/java/com/yahoo/elide/datastores/search/FilterExpressionToLuceneQuery.java]: `return builder.bool() ...`

for consistancy

File: elide-datastore/elide-datastore-search/src/main/java/com/yahoo/elide/datastores/search/SearchDataTransaction.java [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+            List<Object[]> results = fullTextQuery
+                    .setProjection(ProjectionConstants.THIS)
+                    .getResultList();
+
+            if (pagination.isPresent() && pagination.get().isGenerateTotals()) {
+                pagination.get().setPageTotals(fullTextQuery.getResultSize());
+            }
+
+            if (results.isEmpty()) {
+                return new ArrayList<>();
```
Comment (jkusa) [inline elide-datastore/elide-datastore-search/src/main/java/com/yahoo/elide/datastores/search/SearchDataTransaction.java]:
```java
Collections.emptyList();
```

File: elide-datastore/elide-datastore-search/src/main/java/com/yahoo/elide/datastores/search/SearchDataStore.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+   public void populateEntityDictionary(EntityDictionary entityDictionary) {
+       wrapped.populateEntityDictionary(entityDictionary);
+
+       if (indexOnStartup) {
+
+           FullTextEntityManager em = Search.getFullTextEntityManager(entityManagerFactory.createEntityManager());
+           try {
+               for (Class<?> entityClass : entityDictionary.getBindings()) {
+                   if (entityDictionary.getAnnotation(entityClass, Indexed.class) != null) {
+                       em.createIndexer(entityClass).startAndWait();
```
Comment (jkusa) [inline elide-datastore/elide-datastore-search/src/main/java/com/yahoo/elide/datastores/search/SearchDataStore.java]: What is the behavior of waiting? Will elide return a 503 Service Unavailable when waiting?

File: elide-datastore/elide-datastore-search/src/test/java/com/yahoo/elide/datastores/search/SearchDataStoreTest.java [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+        verify(wrappedTransaction, never()).loadObjects(any(), any(), any(), any(), any());
+    }
+
+    @Test
+    public void testContainsPredicate() throws Exception {
+
+        DataStoreTransaction testTransaction = searchStore.beginReadTransaction();
+
+        //Case insensitive query against case insensitive index
+        FilterExpression filter = filterParser.parseFilterExpression("name==*DrU*", Item.class, false);
```
Comment (jkusa) [inline elide-datastore/elide-datastore-search/src/test/java/com/yahoo/elide/datastores/search/SearchDataStoreTest.java]: Can we add a test that verifies you can pass through an escaped value?

File: elide-datastore/elide-datastore-search/README.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+```java
+@Entity
+@Include
+@Indexed
+@Data
+@AnalyzerDef(name = "case_sensitive",
+        tokenizer = @TokenizerDef(factory = StandardTokenizerFactory.class)
+)
+@AnalyzerDef(name = "case_insensitive",
```
Comment (jkusa) [inline elide-datastore/elide-datastore-search/README.md]: Can we mention that these annotations can be at the `package-info.java`  to remove the boilerplate?

File: elide-datastore/elide-datastore-search/src/test/java/com/yahoo/elide/datastores/search/SearchDataStoreTest.java [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+        List<Long> actualIds = StreamSupport.stream(actual.spliterator(), false)
+                .map((obj) -> (Item) obj)
+                .map(Item::getId)
+                .sorted()
+                .collect(Collectors.toList());
+
+        List<Long> expectedIdsSorted = expectedIds.stream().sorted().collect(Collectors.toList());
+
+        Assert.assertEquals(actualIds, expectedIdsSorted);
+    }
```
Comment (jkusa) [inline elide-datastore/elide-datastore-search/src/test/java/com/yahoo/elide/datastores/search/SearchDataStoreTest.java]: Are there any negative tests that can be added?

File: elide-datastore/elide-datastore-search/README.md (hunk 3 above)
Comment (aklish) [reply elide-datastore/elide-datastore-search/README.md]: `@Data` is lombok

File: elide-datastore/elide-datastore-search/src/main/java/com/yahoo/elide/datastores/search/SearchDataStore.java (hunk 7 above)
Comment (aklish) [reply elide-datastore/elide-datastore-search/src/main/java/com/yahoo/elide/datastores/search/SearchDataStore.java]: Assuming you are using elide-standalone (Jetty), this happens inside `server.start()`.  This means the server hasn't yet opened a port for new connections.

The client would get a connection refused error:
```Failed to connect to localhost port 8080: Connection refused```

File: elide-datastore/elide-datastore-search/README.md (hunk 4 above)
Comment (aklish) [reply elide-datastore/elide-datastore-search/README.md]: 👍

File: elide-datastore/elide-datastore-search/src/main/java/com/yahoo/elide/datastores/search/FilterExpressionToLuceneQuery.java (hunk 5 above)
Comment (aklish) [reply elide-datastore/elide-datastore-search/src/main/java/com/yahoo/elide/datastores/search/FilterExpressionToLuceneQuery.java]: 👍

File: elide-datastore/elide-datastore-search/src/main/java/com/yahoo/elide/datastores/search/SearchDataTransaction.java (hunk 6 above)
Comment (aklish) [reply elide-datastore/elide-datastore-search/src/main/java/com/yahoo/elide/datastores/search/SearchDataTransaction.java]: 👍

File: elide-datastore/elide-datastore-search/README.md (hunk 9 above)
Comment (aklish) [reply elide-datastore/elide-datastore-search/README.md]: 👍

File: elide-core/src/main/java/com/yahoo/elide/core/filter/dialect/RSQLFilterDialect.java [hunk 11]
```diff
@@ -157,21 +151,36 @@ public FilterExpression parseGlobalExpression(String path, MultivaluedMap<String
...
             }
         }
         return expressionByType;
     }
 
+
+    /**
+     * Parses a RSQL string into an Elide FilterExpression.
+     * @param expressionText the RSQL string
+     * @param entityType The type associated with the predicate
```
Comment (jkusa) [inline elide-core/src/main/java/com/yahoo/elide/core/filter/dialect/RSQLFilterDialect.java]: Can we add `allowNestedToManyAssociations`

File: elide-datastore/elide-datastore-search/src/test/java/com/yahoo/elide/datastores/search/DataStoreLoadTest.java [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+
+        DataStore mockStore = mock(DataStore.class);
+        wrappedTransaction = mock(DataStoreTransaction.class);
+        when(mockStore.beginReadTransaction()).thenReturn(wrappedTransaction);
+
+        EntityManagerFactory emf = Persistence.createEntityManagerFactory("searchDataStoreTest");
+
+        searchStore = new SearchDataStore(mockStore, emf, true, 3, 10);
+        searchStore.populateEntityDictionary(dictionary);
+
```
Comment (jkusa) [inline elide-datastore/elide-datastore-search/src/test/java/com/yahoo/elide/datastores/search/DataStoreLoadTest.java]: extra `\n`

**Pull request merged**